libraries/Microsoft.Bot.Builder.Core.Extensions/CatchExceptionMiddleware.cs

```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    public class CatchExceptionMiddleware : IMiddleware
+    {
+        private readonly CatchExceptionHandler _handler;
+
+        public CatchExceptionMiddleware(CatchExceptionHandler handler)
+        {
+            _handler = handler;
+        }
+
+        public async Task OnProcessRequest(IBotContext context, MiddlewareSet.NextDelegate next)
+        {
+            await CatchError(context, next);
+        }
+
+        private async Task CatchError(IBotContext context, MiddlewareSet.NextDelegate next)
+        {
+            try
+            {
+                await next();
+            }
+            catch (Exception ex)
```

Comment:
Perhaps only certain exception types? Should the caller be able to specify? 

.Use( new ExceptionHandler<InvalidOperationException>( (ex)=> { mycode; } );

Reply:
Now that the CatchExceptionMiddleware takes a generic type of Exception, these can be stacked, like so....

```cs
.Use(new CatchExceptionMiddleware<Exception>((context, exception) =>
                {
                    context.SendActivity(context.Request.CreateReply(exception.Message));
                    return Task.CompletedTask;
                }))
                .Use(new CatchExceptionMiddleware<NullReferenceException>((context, exception) =>
                {
                    context.SendActivity(context.Request.CreateReply("Sorry - Null Reference Exception"));
                    return Task.CompletedTask;
                }));
```